
# 📘 Comprehensive Guide to Pandas `read_excel()`

This notebook is a **complete teaching resource** for:
- Reading Excel files with Pandas
- Handling multiple sheets
- Working with complex Excel structures
- Cleaning data during import
- Performance optimization
- Common errors and debugging

---



## 1️⃣ Introduction to `read_excel()`

`pandas.read_excel()` is used to read Excel files into a Pandas DataFrame.

### Basic Syntax

```python
pd.read_excel(io, sheet_name=0)
```

### Required Import


In [46]:

import pandas as pd



## 2️⃣ Reading a Single Excel Sheet

### Example 1: Read First Sheet (Default)


In [47]:

df = pd.read_excel("sample.xlsx")
df.head()


,OrderID,Customer,Region,Product,Quantity,UnitPrice,OrderDate
0,ORD001,Customer_1,East,Phone,7,420,2025-09-11
1,ORD002,Customer_2,East,Phone,1,712,2025-10-23
2,ORD003,Customer_3,South,Monitor,8,642,2025-11-04
3,ORD004,Customer_4,West,Monitor,6,458,2025-04-21
4,ORD005,Customer_5,West,Phone,7,1384,2025-12-02



### Example 2: Read Specific Sheet by Name


In [48]:

df = pd.read_excel("sample.xlsx", sheet_name="Sales")
df.head()


,OrderID,Customer,Region,Product,Quantity,UnitPrice,OrderDate
0,ORD001,Customer_1,East,Phone,7,420,2025-09-11
1,ORD002,Customer_2,East,Phone,1,712,2025-10-23
2,ORD003,Customer_3,South,Monitor,8,642,2025-11-04
3,ORD004,Customer_4,West,Monitor,6,458,2025-04-21
4,ORD005,Customer_5,West,Phone,7,1384,2025-12-02



### Example 3: Read Sheet by Index


In [49]:

df = pd.read_excel("sample.xlsx", sheet_name=1)
df.head()


,TransactionID,Department,ExpenseType,Amount,TransactionDate
0,TXN001,HR,Training,18046,2025-04-16
1,TXN002,Marketing,Software,14103,2025-01-10
2,TXN003,HR,Software,4942,2025-05-27
3,TXN004,Operations,Training,19794,2025-04-18
4,TXN005,Marketing,Maintenance,7523,2025-02-20



## 3️⃣ Reading Multiple Sheets

When Excel contains multiple sheets.

### Read All Sheets


In [50]:

all_sheets = pd.read_excel("sample.xlsx", sheet_name=None)
type(all_sheets)


dict


`sheet_name=None` returns a dictionary where:
- Keys → Sheet names
- Values → DataFrames


In [51]:

all_sheets.keys()


dict_keys(['Sales', 'Finance', 'Health'])

In [52]:

sales_df = all_sheets["Sales"]
sales_df.head()


,OrderID,Customer,Region,Product,Quantity,UnitPrice,OrderDate
0,ORD001,Customer_1,East,Phone,7,420,2025-09-11
1,ORD002,Customer_2,East,Phone,1,712,2025-10-23
2,ORD003,Customer_3,South,Monitor,8,642,2025-11-04
3,ORD004,Customer_4,West,Monitor,6,458,2025-04-21
4,ORD005,Customer_5,West,Phone,7,1384,2025-12-02



### Read Multiple Selected Sheets


In [53]:

selected_sheets = pd.read_excel("sample.xlsx", sheet_name=["Sales", "Finance"])
selected_sheets["Finance"].head()


,TransactionID,Department,ExpenseType,Amount,TransactionDate
0,TXN001,HR,Training,18046,2025-04-16
1,TXN002,Marketing,Software,14103,2025-01-10
2,TXN003,HR,Software,4942,2025-05-27
3,TXN004,Operations,Training,19794,2025-04-18
4,TXN005,Marketing,Maintenance,7523,2025-02-20



## 4️⃣ Important Parameters of `read_excel()`

### 1. `usecols`
Read specific columns only.


In [54]:

df = pd.read_excel("sample.xlsx", usecols="A:C")
df.head()


,OrderID,Customer,Region
0,ORD001,Customer_1,East
1,ORD002,Customer_2,East
2,ORD003,Customer_3,South
3,ORD004,Customer_4,West
4,ORD005,Customer_5,West



### 2. `skiprows`
Skip unwanted rows.


In [55]:

df = pd.read_excel("sample.xlsx", skiprows=2)
df.head()


,ORD002,Customer_2,East,Phone,1,712,2025-10-23 00:00:00
0,ORD003,Customer_3,South,Monitor,8,642,2025-11-04
1,ORD004,Customer_4,West,Monitor,6,458,2025-04-21
2,ORD005,Customer_5,West,Phone,7,1384,2025-12-02
3,ORD006,Customer_6,West,Monitor,7,539,2025-12-19
4,ORD007,Customer_7,South,Phone,6,1047,2025-07-10



### 3. `nrows`
Limit number of rows.


In [56]:

df = pd.read_excel("sample.xlsx", nrows=10)
df.head()


,OrderID,Customer,Region,Product,Quantity,UnitPrice,OrderDate
0,ORD001,Customer_1,East,Phone,7,420,2025-09-11
1,ORD002,Customer_2,East,Phone,1,712,2025-10-23
2,ORD003,Customer_3,South,Monitor,8,642,2025-11-04
3,ORD004,Customer_4,West,Monitor,6,458,2025-04-21
4,ORD005,Customer_5,West,Phone,7,1384,2025-12-02



### 4. `header`
Define header row.


In [57]:

df = pd.read_excel("sample.xlsx", header=2)
df.head()


,ORD002,Customer_2,East,Phone,1,712,2025-10-23 00:00:00
0,ORD003,Customer_3,South,Monitor,8,642,2025-11-04
1,ORD004,Customer_4,West,Monitor,6,458,2025-04-21
2,ORD005,Customer_5,West,Phone,7,1384,2025-12-02
3,ORD006,Customer_6,West,Monitor,7,539,2025-12-19
4,ORD007,Customer_7,South,Phone,6,1047,2025-07-10



### 5. `index_col`
Set index column.


In [58]:

df = pd.read_excel("sample.xlsx", index_col=0)
df.head()


,Customer,Region,Product,Quantity,UnitPrice,OrderDate
OrderID,,,,,,
ORD001,Customer_1,East,Phone,7,420,2025-09-11
ORD002,Customer_2,East,Phone,1,712,2025-10-23
ORD003,Customer_3,South,Monitor,8,642,2025-11-04
ORD004,Customer_4,West,Monitor,6,458,2025-04-21
ORD005,Customer_5,West,Phone,7,1384,2025-12-02



### 6. `dtype`
Force specific data types.


In [59]:

df = pd.read_excel("sample.xlsx", dtype={"ID": str})
df.dtypes


OrderID              object
Customer             object
Region               object
Product              object
Quantity              int64
UnitPrice             int64
OrderDate    datetime64[ns]
dtype: object


### 7. `parse_dates`
Parse date columns automatically.


What parse_dates Does in pd.read_excel()

parse_dates tells Pandas:

"Convert this column from normal text (object) into real datetime format automatically while reading the file."

Without parse_dates

If you run:

In [60]:
df = pd.read_excel("sample.xlsx", sheet_name="Finance")
df.dtypes

TransactionID              object
Department                 object
ExpenseType                object
Amount                      int64
TransactionDate    datetime64[ns]
dtype: object

That means Pandas is treating the date like ordinary text.

And text dates:

- Cannot be used for time filtering properly

- Cannot extract year/month/day easily

- Cannot do time-based analysis

`   It becomes a true datetime object.  `

In [61]:

df = pd.read_excel("sample.xlsx", parse_dates=["Date"])
df.dtypes # The error is happening because:
          # There is no column named "Date" in sample.xlsx.


ValueError: Missing column provided to 'parse_dates': 'Date' (sheet: 0)

If Reading Sales SheetIf Reading Sales Sheet

In [ ]:
df = pd.read_excel("sample.xlsx", sheet_name="Sales", parse_dates=["OrderDate"])
df.dtypes

OrderID              object
Customer             object
Region               object
Product              object
Quantity              int64
UnitPrice             int64
OrderDate    datetime64[ns]
dtype: object

If Reading Finance Sheet

In [ ]:
df = pd.read_excel("sample.xlsx", sheet_name="Finance", parse_dates=["TransactionDate"])
df.dtypes

TransactionID              object
Department                 object
ExpenseType                object
Amount                      int64
TransactionDate    datetime64[ns]
dtype: object

If Reading Health Sheet

In [ ]:
df = pd.read_excel("sample.xlsx", sheet_name="Health",
                   parse_dates=["AdmissionDate", "DischargeDate"])
df.dtypes

PatientID                object
Age                       int64
Gender                   object
Diagnosis                object
AdmissionDate    datetime64[ns]
DischargeDate    datetime64[ns]
dtype: object

`   NOTE: Before using parse_dates, always check the column names:  `

`   This prevents 80% of beginner errors.    `

Why This Is Powerful

Once converted, you can now do:

In [ ]:
# Extract Year

df["TransactionDate"].dt.year

0     2025
1     2025
2     2025
3     2025
4     2025
5     2025
6     2025
7     2025
8     2025
9     2025
10    2025
11    2025
12    2025
13    2025
14    2025
15    2025
16    2025
17    2025
18    2025
19    2025
20    2025
21    2025
22    2025
23    2025
24    2025
25    2025
26    2025
27    2025
28    2025
29    2025
30    2025
31    2025
32    2025
33    2025
34    2025
35    2025
36    2025
37    2025
38    2025
39    2025
Name: TransactionDate, dtype: int32

In [ ]:
# Extract Month

df["TransactionDate"].dt.month

0      4
1      1
2      5
3      4
4      2
5      3
6      4
7      6
8      7
9      4
10    11
11    10
12    11
13     8
14     4
15     7
16     5
17    11
18     2
19     8
20    11
21     4
22     9
23     8
24     2
25     1
26    11
27     7
28     9
29     1
30     3
31    12
32     6
33    11
34     9
35     5
36    11
37     3
38     4
39     4
Name: TransactionDate, dtype: int32

In [ ]:
# Filter by Date

df[df["TransactionDate"] > "2025-06-01"]

,TransactionID,Department,ExpenseType,Amount,TransactionDate
7,TXN008,Operations,Maintenance,9661,2025-06-02
8,TXN009,HR,Training,15508,2025-07-25
10,TXN011,HR,Training,11371,2025-11-22
11,TXN012,Marketing,Salary,6602,2025-10-23
12,TXN013,HR,Maintenance,4358,2025-11-03
13,TXN014,HR,Software,17221,2025-08-21
15,TXN016,Operations,Salary,17700,2025-07-16
17,TXN018,HR,Salary,1999,2025-11-11
19,TXN020,Operations,Maintenance,1688,2025-08-20
20,TXN021,Marketing,Software,8636,2025-11-12


In [ ]:
# Time-Based Grouping

df.groupby(df["TransactionDate"].dt.month)["Amount"].sum()

TransactionDate
1      39229
2      40127
3      41591
4     122857
5      27936
6      28064
7      45833
8      26507
9      35440
10      6602
11     61291
12     18929
Name: Amount, dtype: int64

We grouped transactions by month and calculated the total amount spent in each month.

What This Code Means

(1) df["TransactionDate"]

Selects the date column.

(2) .dt.month

Extracts the month number from each date.

Example:

TransactionDate	        dt.month
2025-01-15	            1
2025-03-22	            3
2025-11-01	            11

So now Pandas is grouping by month number (1–12).

(3) df.groupby(...)

This tells Pandas:

“Group rows that belong to the same month together.”

(4) ["Amount"]

Select only the Amount column.

(5) .sum()

Add up all the amounts inside each month group.

The left column (1–12) is called the Group Index.

It represents:

- 1 = January

- 2 = February

- 3 = March

- ...

- 12 = December

So:

Month 4 (April) had total expenses of 122,857

Month 11 had total expenses of 61,291


## 5️⃣ Handling Large Excel Files

### Memory Optimization Tips:
- Use `usecols`
- Use `dtype`
- Read only required sheets
- Convert to CSV for very large files



## 6️⃣ Reading Excel from a URL


In [ ]:
url = "https://github.com/prodigy234/My-Datasets/blob/master/airbnb_room_type.xlsx"

df = pd.read_excel(url)

df.head()

df = pd.read_excel(url)

df.head()

ValueError: Excel file format cannot be determined, you must specify an engine manually.


## 7️⃣ Engine Parameter

Pandas supports different Excel engines:

- openpyxl (for .xlsx)
- xlrd (older .xls)
- odf (for .ods)


In [ ]:

df = pd.read_excel("sample.xlsx", engine="openpyxl")
df.head()



## 8️⃣ Common Errors & Fixes

### Error 1: FileNotFoundError
✔ Check file path.

### Error 2: ValueError: Worksheet not found
✔ Confirm sheet name.

### Error 3: Engine errors
✔ Install openpyxl:

```bash
pip install openpyxl
```



## 9️⃣ Classroom Practical Exercises

1. Load an Excel file with 3 sheets.
2. Extract only one sheet.
3. Load only columns A, C, and F.
4. Skip first 3 rows.
5. Convert a Date column to datetime.
6. Load all sheets and merge them.

---

This notebook covers beginner to advanced usage of `pd.read_excel()` and is ready for classroom teaching.
